Add SBS manufactuer access support to fuel gauge API - includes adding write to API

aaronemassey opened the pull request:
This is part of the ongoing fuel gauge effort described in
https://github.com/zephyrproject-rtos/zephyr/issues/44847

In this PR:

* Add `set_property` to fuel gauge API
* Add read/write SBS specific mfr access to fuel gauge API
* Implement/test mfr access read/writing in SBS Gauge driver


The `set_property` added to the API is intended to mirror the `get_property`
API.

Adding a `set_property` in no way implies we're finished with the `get_property`
or adding support for more SBS spec properties to the SBS Gauge Driver
itself. Current strategy is to prototype this API by maintaining a breadth of
the kind of features it supports with the SBS Gauge driver as an example.

aaronemassey committed:
fuel_gauge: Add readable mfr access prop

Add the SBS manufactuer access as a fuel gauge readable property.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey committed:
fuel_gauge: Implement sbs gauge read for mfr acc

Implement and weakly test fetching the manufacturer access word from an SBS
compatible fuel gauge.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey committed:
fuel_gauge: Add writable mfr access prop

Add fuel_gauge_set_property function to the fuel gauge API with starting
writable property of SBS specific manufactuer access.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey committed:
fuel_gauge: sbs_gauge: Enable MFR ACC write

Update the SBS Gauge driver that implements the fuel_gauge API to implement
a set_property function allowing the writing of an SBS word to the
manufacturer access register per the SBS spec.

Includes an update to the SBS Gauge emulator and SBS fuel gauge tests to
weakly verify the code runs.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

alevkoy commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -133,10 +152,26 @@ struct fuel_gauge_get_property {
+ * @param dev Pointer to the battery fuel-gauge device
+ * @param props pointer to array of fuel_gauge_set_property struct where the property struct
+ * field is set by the caller to determine what property is written to the
+ * fuel gauge device from the fuel_gauge_get_property struct's value field.
```
Weird line break?

aaronemassey replied:
:+1:

yperess commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:
```diff
@@ -102,6 +138,18 @@ static int sbs_gauge_get_props(const struct device *dev, struct fuel_gauge_get_p
+		ret |= sbs_gauge_set_prop(dev, props + i);
+	}
+
+	return ret;
```
This will be garbage if we ever return 2 different errors from sbs_gauge_set_prop. Consider replacing with `return ret == 0 ? 0 : -EINVAL;`

alevkoy replied:
I was wondering about that, but I saw that the read function already worked that way, so I figured somebody liked it like that. I'm a bit curious about the nominal positive return values when some reads/writes failed, and others succeeded. I'm not sure what I'd do with that as the caller. How is it supposed to be used?

aaronemassey replied:
Made https://github.com/zephyrproject-rtos/zephyr/pull/52733

alevkoy replied:
I still think this is weird, but #52992 has been opened to that end.

aaronemassey replied:
Yeah I'd like to add the `__subsystem` bit then focus on https://github.com/zephyrproject-rtos/zephyr/issues/52992 to get finalized.

aaronemassey commented:
Looks like some of my emulator logic is a bit buggy wrt I2C messages. Fixing that in another PR and rebasing this onto that one. For now, consider this a rough draft.

aaronemassey committed:
emul: Correct emul_sbs_gauge to allow SBS word r/w

Fix SBS I2C transfer emulation to accept reads and writes of 16 bit words
as defined by the SBS spec. This change is tested by a following commit
that implements writing and subsequently reading a written SBS property.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey commented:
Next PR is to add `__subsystem` and similar piping to make this API actually do syscalls.

keith-zephyr commented on subsys/emul/i2c/emul_sbs_gauge.c:
```diff
@@ -115,17 +115,21 @@ static int sbs_gauge_emul_transfer_i2c(const struct emul *target, struct i2c_msg
 				}
-				msgs->buf[0] = val;
+
+				sys_put_le16(val, msgs->buf);
```
The endian conversion needs a comment. As is, this looks asymmetrical because the read path does the conversion and the write path doesn't.

aaronemassey replied:
Done.

alevkoy commented on subsys/emul/i2c/emul_sbs_gauge.c:
```diff
@@ -114,17 +120,22 @@ static int sbs_gauge_emul_transfer_i2c(const struct emul *target, struct i2c_msg
 				}
-				msgs->buf[0] = val;
+
+				/* SBS uses SMBus which sends data in little-endian format. */
```
Nit:
```
uses SMBus, which sends
          ^
```

aaronemassey replied:
Done.

aaronemassey commented on subsys/emul/i2c/emul_sbs_gauge.c:
```diff
@@ -115,17 +115,21 @@ static int sbs_gauge_emul_transfer_i2c(const struct emul *target, struct i2c_msg
 			}
 		} else {
-			if (msgs->len != 1) {
+			/* We write a word (2 bytes by the SBS spec */
```
Good catch. Done.
